File: ios/App/App/AppOpenAdPlugin.swift
import Foundation
import Capacitor
import GoogleMobileAds

@objc(AppOpenAdPlugin)
public class AppOpenAdPlugin: CAPPlugin, CAPBridgedPlugin, FullScreenContentDelegate {
    public let identifier = "AppOpenAdPlugin"
    public let jsName = "AppOpenAd"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "prepare", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise)
    ]

    private let maximumCacheAge: TimeInterval = 4 * 60 * 60
    private var appOpenAd: AppOpenAd?
    private var loadedAt: Date?
    private var isLoading = false
    private var isShowing = false

    private func isAvailable() -> Bool {
        guard let loadedAt else { return false }
        return appOpenAd != nil && Date().timeIntervalSince(loadedAt) < maximumCacheAge
    }

    @objc func prepare(_ call: CAPPluginCall) {
        guard let adId = call.getString("adId")?.trimmingCharacters(in: .whitespacesAndNewlines),
              !adId.isEmpty else {
            call.reject("App open ad unit is missing", "admob/app-open-id-missing")
            return
        }
        if isAvailable() {
            call.resolve(["ready": true])
            return
        }
        if isLoading {
            call.reject("App open ad is already loading", "admob/app-open-loading")
            return
        }

        isLoading = true
        Task { @MainActor in
            do {
                let ad = try await AppOpenAd.load(with: adId, request: Request())
                self.appOpenAd = ad
                self.appOpenAd?.fullScreenContentDelegate = self
                self.loadedAt = Date()
                self.isLoading = false
                call.resolve(["ready": true])
            } catch {
                self.appOpenAd = nil
                self.loadedAt = nil
                self.isLoading = false
                call.reject(
                    error.localizedDescription,
                    "admob/app-open-load-failed",
                    error
                )
            }
        }
    }

    @objc func show(_ call: CAPPluginCall) {
        if isShowing {
            call.reject("App open ad is already showing", "admob/app-open-showing")
            return
        }
        guard isAvailable(), let ad = appOpenAd else {
            appOpenAd = nil
            loadedAt = nil
            call.reject("App open ad is not ready", "admob/app-open-not-ready")
            return
        }

        DispatchQueue.main.async {
            self.isShowing = true
            ad.present(from: self.getRootVC())
            call.resolve(["shown": true])
        }
    }

    public func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        appOpenAd = nil
        loadedAt = nil
        isShowing = false
        notifyListeners("dismissed", data: [:])
    }

    public func ad(
        _ ad: FullScreenPresentingAd,
        didFailToPresentFullScreenContentWithError error: Error
    ) {
        appOpenAd = nil
        loadedAt = nil
        isShowing = false
        notifyListeners("failedToShow", data: [
            "message": error.localizedDescription
        ])
    }

    public func adWillPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
        notifyListeners("showed", data: [:])
    }
}

File: android/app/src/main/java/com/fortunecookieai/app/AppOpenAdPlugin.java
package com.fortunecookieai.app;

import androidx.annotation.NonNull;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.android.gms.ads.AdError;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.FullScreenContentCallback;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.appopen.AppOpenAd;

@CapacitorPlugin(name = "AppOpenAd")
public class AppOpenAdPlugin extends Plugin {
    private static final long MAX_CACHE_AGE_MS = 4L * 60L * 60L * 1000L;

    private AppOpenAd appOpenAd;
    private long loadedAt;
    private boolean loading;
    private boolean showing;

    private boolean isAvailable() {
        return appOpenAd != null && (System.currentTimeMillis() - loadedAt) < MAX_CACHE_AGE_MS;
    }

    @PluginMethod
    public void prepare(PluginCall call) {
        final String adId = call.getString("adId", "").trim();
        if (adId.isEmpty()) {
            call.reject("App open ad unit is missing", "admob/app-open-id-missing");
            return;
        }
        if (isAvailable()) {
            JSObject result = new JSObject();
            result.put("ready", true);
            call.resolve(result);
            return;
        }
        if (loading) {
            call.reject("App open ad is already loading", "admob/app-open-loading");
            return;
        }

        loading = true;
        getActivity().runOnUiThread(() -> AppOpenAd.load(
            getContext(),
            adId,
            new AdRequest.Builder().build(),
            new AppOpenAd.AppOpenAdLoadCallback() {
                @Override
                public void onAdLoaded(@NonNull AppOpenAd ad) {
                    appOpenAd = ad;
                    loadedAt = System.currentTimeMillis();
                    loading = false;
                    JSObject result = new JSObject();
                    result.put("ready", true);
                    call.resolve(result);
                }

                @Override
                public void onAdFailedToLoad(@NonNull LoadAdError error) {
                    appOpenAd = null;
                    loadedAt = 0L;
                    loading = false;
                    call.reject(
                        error.getMessage(),
                        "admob/app-open-load-failed"
                    );
                }
            }
        ));
    }

    @PluginMethod
    public void show(PluginCall call) {
        if (showing) {
            call.reject("App open ad is already showing", "admob/app-open-showing");
            return;
        }
        if (!isAvailable()) {
            appOpenAd = null;
            call.reject("App open ad is not ready", "admob/app-open-not-ready");
            return;
        }

        final AppOpenAd ad = appOpenAd;
        showing = true;
        ad.setFullScreenContentCallback(new FullScreenContentCallback() {
            @Override
            public void onAdDismissedFullScreenContent() {
                appOpenAd = null;
                loadedAt = 0L;
                showing = false;
                notifyListeners("dismissed", new JSObject());
            }

            @Override
            public void onAdFailedToShowFullScreenContent(@NonNull AdError error) {
                appOpenAd = null;
                loadedAt = 0L;
                showing = false;
                JSObject data = new JSObject();
                data.put("message", error.getMessage());
                notifyListeners("failedToShow", data);
            }

            @Override
            public void onAdShowedFullScreenContent() {
                notifyListeners("showed", new JSObject());
            }
        });

        getActivity().runOnUiThread(() -> {
            ad.show(getActivity());
            JSObject result = new JSObject();
            result.put("shown", true);
            call.resolve(result);
        });
    }
}
